File: docs/examples/file-browser/file-browser.py
#!/usr/bin/env python3

import json
import pathlib
import argparse
import sys

parser = argparse.ArgumentParser()
parser.add_argument("root", nargs="?", type=str)
parser.add_argument("--show-hidden", action="store_true")
args = parser.parse_args()

root: pathlib.Path = pathlib.Path(args.root) if args.root else pathlib.Path.cwd()
entries = root.iterdir()
if not args.show_hidden:
    entries = filter(lambda p: not p.name.startswith("."), entries)

json.dump(
    {
        "type": "list",
        "items": [
            {
                "title": path.name,
                "accessories": [str(root.absolute())],
                "actions": [
                    (
                        {
                            "type": "open",
                            "title": "Open File",
                            "target": f"file://{path.absolute()}",
                        }
                        if path.is_file()
                        else {
                            "type": "run",
                            "onSuccess": "push",
                            "title": "Browse Directory",
                            "command": f"{sys.argv[0]} {path.absolute()}",
                        }
                    ),
                    {
                        "type": "copy",
                        "title": "Copy Path",
                        "shortcut": "ctrl+y",
                        "text": str(path.absolute()),
                    },
                ],
            }
            for path in sorted(entries, key=lambda p: p.name)
        ],
    },
    sys.stdout,
)

File: docs/examples/file-browser/file-browser.ts
#!/usr/bin/env -S deno run --allow-read

import * as sunbeam from "../../../pkg/typescript/mod.ts";
import * as path from "https://deno.land/std@0.201.0/path/mod.ts";
import { handler } from "../../../pkg/typescript/http.ts";

const extension = new sunbeam.Extension({
  title: "File Browser",
  description: "A file browser extension for Sunbeam",
}).command({
  name: "browse",
  title: "Browse files",
  mode: "filter",
  params: [
    {
      name: "root",
      type: "string",
      optional: true,
    },
  ],
  run: async ({ params }) => {
    const root = params.root as string || Deno.cwd();

    const entries = [];
    for await (const entry of Deno.readDir(root)) {
      entries.push(entry);
    }

    const items: sunbeam.Listitem[] = entries.map((entry) => {
      const actions: sunbeam.Action[] = [];
      const filepath = path.join(root, entry.name);

      if (entry.isDirectory) {
        actions.push({
          title: "Browse Directory",
          type: "run",
          command: {
            name: "browse",
            params: {
              root: path.join(root, entry.name),
            },
          },
        });
      }

      return {
        title: entry.name,
        subtitle: filepath,
        actions,
      };
    });

    return {
      title: `${root}`,
      items,
    };
  },
});

Deno.serve(handler(extension));
